GetStream/stream-chat-android
# Feature/3929 bidirectional pagination

### 🎯 Goal

Rerevert of #3818 

### 🛠 Implementation details

Updated `ChannelLogic` which determines if the pagination has reached either end or start of the messages list. Also added checks when receiving or sending new messages, so that the new message isn't added to the list if the newest messages are not loaded to avoid breaking pagination of newer messages.

Introduced new pagination handler for reaching the end of new messages to `MessageList` composable. Added scroll to bottom handler to `MessagesList`.

### 🎨 UI Changes

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/38032787/176719148-72f40164-bf9b-433e-9f5e-190f44cdc6ea.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/38032787/176719135-254b3323-06c0-4064-9f9a-c2bba4cf9157.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing

1. Build this branch and enter a channel that has more than one page
2. Scroll n number of pages and reply to a message
4. Exit the channel and enter it again, the message history will be reset
5. Click on the quoted message which should load the message and a page that surrounds the message
6. Scroll up and down to ensure that the bi directional pagination is working
7. Try sending messages from the current user and from a different device, messages shouldn't be added to the bottom of the list, otherwise pagination will break
8. Try adding reactions to messages you are scrolled to, replying to them and deleting them to insure that the messages are properly updated
9. Click on scroll to bottom button, if the newest messages aren't present they should be loaded and the list should scroll to the bottom
10. Background the app when newest messages are not loaded and send new messages from another device, the messages should not be added to the list.

### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist

### 🎉 GIF

![giphy](https://user-images.githubusercontent.com/38032787/185102923-9660c1fa-ac34-43f7-9ced-e78d6f39db69.gif)
